tmt/utils/git.py

```diff
@@ -26,6 +27,114 @@
     import tmt.base
 
 
+@dataclasses.dataclass
+class GitInfo:
+    """ Data container for commonly queried git data. """
+
+    git_root: Path
+    """Path to the git root."""
+    ref: str
+    """Most human-readable git ref."""
+    remote: str
+    """Git remote linked to the current git ref."""
+    default_branch: Optional[str]
+    """Default branch of the remote."""
+    url: Optional[str]
+    """Public url of the remote."""
+
+    @classmethod
+    @functools.cache
+    def from_fmf_root(cls, *, fmf_root: Path, logger: tmt.log.Logger) -> Optional["GitInfo"]:
+        """
+        Get the current git info of a fmf tree.
+
+        :param fmf_root: Root path of the fmf tree
+        :param logger: Current tmt logger
+        :return: Git info container or ``None`` if git metadata could not be resolved
+        """
+
+        def run(command: Command) -> str:
+            """
+            Run command, return output.
+            We don't need the stderr here, but we need exit status.
+            """
+            result = command.run(cwd=fmf_root, logger=logger)
+            if result.stdout is None:
+                return ""
+            return result.stdout.strip()
+
+        # Prepare url (for now handle just the most common schemas)
+        try:
+            # Check if we are a git repo
+            run(Command("git", "rev-parse", "--is-inside-work-tree"))
+
+            # Initialize common git facts
+            # Get some basic git references for HEAD
+            all_refs = run(Command("git", "for-each-ref", "--format=%(refname)", "--points-at=@"))
+            logger.debug("git all_refs", all_refs)
+            # curr_ref is either HEAD or fully-qualified (branch) reference
+            curr_ref = run(Command("git", "rev-parse", "--symbolic-full-name", "@"))
+            logger.debug("git initial curr_ref", curr_ref)
+            # Get the top-level git_root
+            _git_root = git_root(fmf_root=fmf_root, logger=logger)
+            assert _git_root is not None  # narrow type
+        except RunError:
+            # Not a git repo, everything should be pointing to None at this point
+            return None
+
+        if curr_ref != "HEAD":
+            # The reference is fully qualified -> we are on a branch
+            # Get the short name
+            branch = run(Command("git", "for-each-ref", "--format=%(refname:short)", curr_ref))
+            ref = branch
+        else:
+            # Not on a branch, check if we are on a tag or just a refs
+            try:
+                tags = run(Command("git", "describe", "--tags"))
```

Comment:
`git describe --tags` returns only one tag for me, so the splitlines() is probably not doing much?
fwiw, `git describe --tags --exact-match` can tell if the current commit is tagged.

Reply:
Probably true? I think there is some shenanigans when you are a few commits away from a tag and/or on a branch with different tags, e.g. a release branch. Any of them should be valid as a commitish though.